File: src/binary-search/BOJ10816.ts
export {};
// https://www.acmicpc.net/problem/10816
// 숫자 카드2

const inputStrings = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");

const cards = inputStrings[1]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);

const testCases = inputStrings[3].split(" ").map(Number);
let answer = "";
for (let i = 0; i < testCases.length; i++) {
  answer += solution(testCases[i]);
}
console.log(answer);

function solution(target: number) {
  return `${upperIndex(target) - lowerIndex(target)}\n`;
}

function lowerIndex(target: number) {
  let start = 0;
  let end = cards.length;

  while (start < end) {
    const middle = Math.floor((start + end) / 2);
    if (cards[middle] >= target) {
      end = middle;
    } else {
      start = middle + 1;
    }
  }
  return start;
}

function upperIndex(target: number) {
  let start = 0;
  let end = cards.length;

  while (start < end) {
    const middle = Math.floor((start + end) / 2);
    if (cards[middle] > target) {
      end = middle;
    } else {
      start = middle + 1;
    }
  }

  return start;
}

File: src/binary-search/BOJ10816.py
# https://www.acmicpc.net/problem/10816
# 숫자 카드2

import sys
input = sys.stdin.readline

input()
numbers = list(map(int, input().split()))
numbers.sort()
input()
targets = list(map(int, input().split()))


def solution():
    answers = []
    for target in targets:
        start = 0
        end = len(numbers)
        answer = getLastIdx(start, end, target) - \
            getStartIdx(start, end, target)
        answers.append(str(answer))
    print(" ".join(answers))


def getStartIdx(start, end, target):
    while start < end:
        mid = (start+end) // 2
        if numbers[mid] >= target:
            end = mid
        else:
            start = mid + 1

    return start


def getLastIdx(start, end, target):
    while start < end:
        mid = (start+end) // 2
        if numbers[mid] > target:
            end = mid
        else:
            start = mid + 1
    return start


solution()
